### PR TITLE
Update and move instructions to read me and remove unnecessary requirements

### PR DESCRIPTION
**Pull request by shred86**
Updated the macOS installation instructions and removed the previous used install_mac.sh file since it doesn't really make the process any easier. Additionally, removed some unnecessary requirements as they're automatically installed with pip and not necessarily a requirement for Ortho4XP.

I think it would be better to move these instructions to the main read me file and use markdown for better formatting. More than happy to work on that.

### CODE REVIEW

**Comment by w8sl** [comment]
Alternatively, keep requirements.txt, change directory to Ortho4XP-master and run:
pip3.11 install -r requirements.txt
scikit_fmm-2024.5.29 is the latest version, but it is better to leave it unpinned, as it may be updated soon.
It may be necessary to pin numpy to 1.26.4, as we don´t know if numpy 2.0 will work. 
**pip3.11 install gdal==$(gdal-config --version)**  is required, as only gdal for python 3.12 is installed by brew.

**Comment by shred86** [comment]
I do like `pip3.11 install -r requirements.txt` better. I just wasn't sure if the requirements.txt being proposed in this PR will be accepted since it doesn't contain extra dependencies that pip installs. I suppose those dependencies will get installed regardless so it may be a moot point, but they probably shouldn't be pinned versions in that case (the pip dependencies) or else you may end up reverting to an older version if pip is updated to use a newer version.

As for scikit_fmm, isn't that the point of pinning it to a version? So we can ensure it will work with Ortho4XP? Like the other dependencies, if they're automatically updated and we're just pulling the latest version, there's a chance it could break Ortho4XP.

**Comment by shred86** [comment]
I moved the installation instructions to the main read me which you can preview in [my forked repo](https://github.com/shred86/Ortho4XP/tree/updates). The Window and Linux installation instructions need to be updated. I removed some stuff that is not currently applicable (such as no .exe currently exists for 1.40). The macOS instructions are up-to-date.

**Comment by shred86** [comment]
@ColinM9991 and @telephono tagging you guys in this PR because I just realized you also made a [PR](https://github.com/oscarpilote/Ortho4XP/pull/254) earlier trying to help update and clean up the documentation. I think we could just consolidate it all into one if that works for you guys. Some notes:

- I didn't include using a `venv` in the Windows instructions. I think we should probably just avoid that to reduce confusion and complications for users. I think if you understand and use virtual environments already, then you'll probably just be able to do that on your own.
- I wasn't sure if for Windows, you can just use `python3` or if you have to specify `python3.11`. I'm assuming just `python3` works because you're installing Python through it's official installation package.
- Still need to update the Linux instructions.

**Comment by telephono** [comment]
For Windows it's probably easier to work with Python 3.10
The download site mentioned in the old instructions for the compiled Python packages for Windows no longer exists?

I found the easiest way to get GDAL for Windows is to use the packages from the [GISInternals Support Site](https://www.gisinternals.com/release.php). But they only provide Python bindings up to Python 3.10.

All the other packages are easy to get with pip install ...

**Comment by w8sl** [comment]
Regarding gdal:  try a tile with USGS NED1/3 elevation data on Python 3.11 install.
Usually **pip3.11 install gdal** works, but there are days when PyPI attempts to install different version of gdal than brew 
Therefore **pip3.11 install gdal==$(gdal-config --version)**  is better. Affects only macOS.

Windows:
For some reason it is difficult to find, using internet search, but Christoph Gohlke has moved to GitHub: https://github.com/cgohlke/geospatial-wheels/releases

Latest relase:
https://github.com/cgohlke/geospatial-wheels/releases/download/v2024.2.18/GDAL-3.8.4-cp312-cp312-win_amd64.whl

Also scikit-fmm  2024.5.29  is available for Python 3.12. I don't know which C-compiler is required for Python 12 install. Visual Studio? Cannot verify just now.

Suggested, modified  instructions (tested) for Windows and Python 3.11 (same requirements as for macOS, no need to separately install urrlib, certifi, charsen-normalizer, idna):

1) Download and install Python 3 from www.python.org

Just select one for your Windows OS, there is no benefit in our case to
download the pretty lastest version of Python, since you might get difficulties
further down to find modules already built for it. As of 04/2024, Python 3.11. is recommended.

During the installation, check the following options:

"Use admin privileges when installing py.exe"
"Add python.exe to PATH"
->Customize installation, Optional Features:
"pip"
"tcl/tk and IDLE"
"py launcher"
"for all users"
->Next, Advanced Options:
"Install Python 3.11 for all users"
"Add Python to environment variables"
"Precompile standard library"


2) Download binary wheel for GDAL from https://github.com/cgohlke/geospatial-wheels/releases

Pay attention to take the one that corresponds to the Python version which you picked
at Step 1) and to your OS nbr of bits (32 or 64, I guess 64 in all but a few cases).
As an example, if Python 3.11.*  was selected at Step 1) above and you
have a 64bit windows, then you would choose file that contains -cp311- and _amd64
within filename.


Download and install The Microsoft Visual C++ Redistributable packages for Visual Studio 2022:

https://learn.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170


3) Open Command Prompt window as a standard user

4) Change directory to Ortho4XP folder (freshly downloaded from Github) and launch:

pip install -r requirements.txt

4) Open Command Prompt window as administrator and execute the following command:

pip install /path to GDAL-***.whl file, downloaded at Step 2)

In order to use the built-in GeoTIFF feature of the custom_zl map or custom_dem source you will need to
add the directory containing gdal.py, gdal_translate and gdalwarp (***/python**/lib/site-packages/osgeo/)
into the system PATH variable.

You should be done. Open Command Prompt window in the Ortho4XP directory 
and launch "python Ortho4XP.py".

**Comment by shred86** [comment]
When I try `pip3.11 install gdal==$(gdal-config --version)`, I get:

````
zsh: command not found: gdal-config
ERROR: Could not find a version that satisfies the requirement gdal== (from versions: 1.5.0, 1.5.2, 1.6.0, 1.6.1, 1.7.0, 1.7.1, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.11.0, 1.11.1, 1.11.2, 2.0.0, 2.0.1, 2.1.0, 2.1.3, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.4.0, 2.4.2, 2.4.3, 2.4.4, 3.0.0, 3.0.1, 3.0.2, 3.0.3, 3.0.4, 3.1.0, 3.1.1, 3.1.2, 3.1.3, 3.1.4, 3.2.0, 3.2.1, 3.2.2, 3.2.2.1, 3.2.3, 3.3.0, 3.3.1, 3.3.2, 3.3.3, 3.4.0, 3.4.1, 3.4.2, 3.4.3, 3.5.0, 3.5.0.3, 3.5.1, 3.5.2, 3.5.3, 3.6.0, 3.6.0.1, 3.6.1, 3.6.2, 3.6.3, 3.6.4, 3.7.0, 3.7.1, 3.7.1.1, 3.7.2, 3.7.3, 3.8.0, 3.8.1, 3.8.2, 3.8.3, 3.8.4, 3.8.5, 3.9.0)
ERROR: No matching distribution found for gdal==
````

I did `brew install gdal` originally and it looks like it installed 3.9.0.

**Comment by w8sl** [comment]
It worked before. Problem with Homebrew is that they are often changing gdal dependencies and there is no official way to pin versions. pip3.11 gdal should work - most of the time.

**Comment by shred86** [comment]
I'm assuming we can't put `gdal==$(gdal-config --version)` in the requirements.txt so it will just have to be an additional step after doing `pip3.11 install -r requirements.txt` - is that correct? For my own edification, I'm assuming gdal and the gdal python package are only required if you're using "custom_dem" data other than the default.

As for the Windows instructions, I'm testing them on my Windows laptop now. One issue I ran into is I already have Python 3.12 installed, so when you try to use `python` or `pip`, it's set to Python 3.12 (using `python3.11` and `pip3.11` doesn't seem to work like it does on macOS). Instead, I'm using `py -3.11 -m pip install -r requirements.txt`. I don't see any reason not to just use `py -3.11` in the instructions if Python 3.11 is going to be the installed version - anything I'm missing there? Just helps prevent issues for users and should work even if Python 3.11 is the only version you have installed.

@telephono Why is it easier to work with Python 3.10 on Windows? Appreciate the inputs and feedback.

**Comment by shred86** [comment]
@w8sl I'm incorporating your inputs into the steps for Windows and I'm going through them on my Windows laptop just to make sure it works correctly. You mentioned:

"In order to use the built-in GeoTIFF feature of the custom_zl map or custom_dem source you will need to
add the directory containing gdal.py, gdal_translate and gdalwarp (*/python/lib/site-packages/osgeo/)
into the system PATH variable."

I didn't do that but I'm still able to use NED 1" and 1/3" no problem. Is that step perhaps no longer needed?

**Comment by w8sl** [comment]
It was from the original instructions and only system PATH variable has been working for me, but I am not a Windows expert. What happens, when you just write gdalwarp in Command Prompt widow? If no error, then all is ok.

**Comment by shred86** [comment]
`gdalwarp` from the command prompt returns that it's not recognized. But like I mentioned, I'm still able to build tiles using NED 1" and NED 1/3" with no issues. Maybe something changed in Ortho4XP 1.4 that doesn't need that PATH variable? Or is that step perhaps needed for another Ortho4XP function?

Anyways, I pushed the latest changes on the read me which you can view [here](https://github.com/shred86/Ortho4XP/tree/updates).

**Comment by w8sl** [comment]
It looks good!
Another way to test gdal is to start python session: python enter
then type: "from osgeo import gdal_array"

It must work for NED1/3.
But gdal-free environment is a good starter!

**Comment by shred86** [comment]
Ah, I see now. So we just need it for the [O4_DEM_Utils.py](https://github.com/oscarpilote/Ortho4XP/blob/c5e0d4b50774de1850246fac56b9d461ae026a72/src/O4_DEM_Utils.py#L12) to be able to import. If that's the case, we certainly don't need that step about adding it to the PATH variables. 

I did start start the Python 3.11 interpreter and was able to import `gdal_array` from the `osgeo` library.

As far as I can tell, the Windows steps should be good to go. However, I haven't messed with any of the other functions in Ortho4XP other than custom_dem.

As for macOS, one strange thing is I tried creating a virtual env to just test the steps again but when I try to `pip3.11 install -r requirements.txt` from within the venv, I was getting some error (forgot to save it). But if I install each dependency individually, it works just fine and I can run Ortho4XP.... I then deleted the venv and tried again, and now it works. 🤷 Other then that, I think the macOS steps are good to go with the same caveat I mentioned above about not having tested every function in Ortho4XP.

**Comment by shred86** [comment]
On a second thought, I actually think doing a shell script is a better option. Reason:

- Isolate everything to a Python virtual environment
- Use the latest Python 3.12 version
- Less steps to execute

The process would simply be:

1. Install homebrew
2. chmod 755 ./install_mac.sh
3. Run ./install_mac.sh

Then to run Ortho4XP:

1. Run ./start_Ortho4XP


I tested this tonight by uninstalling everything and setting it back up using this:

````
echo "Setting up Ortho4XP...."
echo " "

# Install software dependencies from brew
brew install python python-tk spatialindex p7zip proj gdal

# Create a Python virtual environment
python3 -m venv venv

# Activate the Python virtual environment
source venv/bin/activate

# Install Python dependencies
pip install -r requirements.txt

# Allow macOS to run required tools
xattr -dr com.apple.quarantine ./Utils/Mac/*

# Make start_mac.sh file executable
chmod 755 start_mac.sh

echo " "
echo "Ortho4XP setup complete!"
````

**Comment by w8sl** [comment]
I am working on script derived from PR #248

**Comment by shred86** [comment]
Ah nice! Some thoughts - Any reason you're [searching for packages](https://github.com/w8sl/Ortho4XP/blob/f425fcd98fbdd470fc9743c28531b98252d68536/z_Install_O4XP_Python_Venv.sh#L41) and then running the install command? Why not just let the package manager do that? I'm just a fan of keeping things simple.

The example I provided above basically does the same thing for macOS with significantly less lines. I do like the check for checking you're in the correct directory and also that homebrew is installed so you're showing meaningful messages.

**Comment by w8sl** [comment]
I know. Big effort is done to avoid brew auto update - if everything works.

**Comment by w8sl** [comment]
My script seems to work on macOS, Ubuntu, Arch. It may be further modified in both ways: simplify or add installation of Homebrew + path to the script.

**Comment by shred86** [comment]
Looking at the commit history, I think it might be a long time before we get anything merged. I decided to just create a fork for now for two reasons.

1. Documentation is lacking and information on Ortho4XP is very fragmented. It took me far too much time and effort to figure out how install and use Ortho4XP (still figuring it out) so I figured I might as well spend more time creating better documentation. 😆 I'm using the Wiki in my forked repo to do this.
2. I realized there's a lot of people with their own small modifications to Ortho4XP that will probably not get merged into this main repo for a while, if ever. So in the meantime, I figure consolidating those updates into one place might be easier for users. 

Another solution is if we can get a couple of us as maintainers to this repo, I'm more than happy to just move the Wiki here and help update/maintain it. Additionally, we can just create a separate `dev` branch here to incorporate changes and leave `master` for when @oscarpilote has time to review. This way, we have the official unmodified version and a more up-to-date version (dev) that people can use.

In the meantime, I'm just going to keep updating my forked repo. Maybe when @oscarpilote hops back on, we can see if the proposed solution above would be something he would entertain. I much prefer trying to keep everything in one location. Feel free to check out the [forked repo](https://github.com/shred86/Ortho4XP/tree/dev), especially the [Wiki](https://github.com/shred86/Ortho4XP/wiki) where I've spent the most time. I'd love to incorporate any and all of the work that you (and others) have done as well.

**Comment by w8sl** [comment]
Great work! Let Oscar decide what is going to the master branch. Wiki is a good idea. Many forks were created in a period of silence.Big changes are merged now to 1.40 and more are promised. The problem with forks is that each one seems to have like 1-5 users. Anyway, it may be a good idea to keep some less popular features in forks in order to minimize changes merged to master.

**Comment by ColinM9991** [comment]
Hi both,

Sorry for not replying sooner. It looks like you've both cracked on with this and developed a decent solution.

I play X-Plane on Windows so I'm more than happy to closey draft PR, pull this down and test Ortho4XP to provide some feedback if this hasn't already been checked. Microsoft have butchered Python management via the MS Store so I'll clean my desktop down first and reinstall 3.11

File: README.md
Comment (ColinM9991) [inline README.md:58]: Tested this locally. Shouldn't this be

```
py -3.11 -m pip install "path_to_downloaded_file_in_step_2/GDAL-3.8.4-cp311-cp311-win_amd64.whl"
```

File: README.md
Comment (ColinM9991) [inline README.md:49]: `Ortho4XP_v130.py` doesn't exist.

Try this instead. Since `Ortho4XP.py` is a script and not a module, the `-m` switch can be removed.
`py -3.11 Ortho4XP.py`

**Review by ColinM9991 — CHANGES_REQUESTED** [review]
Two minor points as I'm going through and testing this on Windows

File: README.md
Comment (shred86) [reply README.md:58]: Yeah that's my bad - I rushed through this and didn't really work on it. I'm going to post another message here with some working steps for Python 3.12.3 on Windows, as well as some more ideas.

**Comment by shred86** [comment]
So I just spent some time on my Windows laptop figuring out how to get things working with the latest version of Python (just to match the macOS and Linux instructions). I've ran through these twice, uninstalling everything on my computer (Python, C++ Build Tools, etc.) and going through the process again and it seems to work.

First idea though, why don't we just include GDAL-3.8.4-cp312-cp312-win_amd64.whl with Ortho4XP? It can go in the Utils/win folder just to simply things (it' s 32MB). We're going to want to pin dependency versions anyways and this is just one less step.

As for the steps I used on Windows 12:

1. Install Python 3.12.3 and make sure "pip", "tcl/tk and IDLE" are selected on the Optional Features page as well as "Add python to environmental variables" on the following page.
2. Download and install [Microsoft C++ Build Tools](https://visualstudio.microsoft.com/visual-cpp-build-tools/). Select "Desktop development with C++" during the installation process. This is required for scikit-fmm.
3. Open a Command Prompt and navigate to the Ortho4XP directory.
4. Create a Python virtual environment: ```python -m venv venv```
5. Activate the Python virtual environment: ```venv\Scripts\activate.bat```
6. Install gdal: ```pip install Utils\win\GDAL-3.8.4-cp312-cp312-win_amd64.whl``` -> realize for this step I had the GDAL wheel file included in my Ortho4XP\Utils\win folder so you'll have to [download](https://github.com/cgohlke/geospatial-wheels/releases/download/v2024.2.18/GDAL-3.8.4-cp312-cp312-win_amd64.whl) it.
7. Install remaining dependencies: ```pip install -r requirements.txt```
8. Start Ortho4XP: ```python Ortho4XP.py```

Step 4 through 7 can all be automated with a .bat script, although I'm having issues getting the .bat script to work. I thought I could do:

`call venv\Scripts\activate.bat & pip install Utils\win\GDAL-3.8.4-cp312-cp312-win_amd64.whl & pip install -r requirements.txt`

And it works if I run it directly from the Command Prompt but if I put it in a .bat file and run it, it stops after the venv is activated. :shrug:

Anyways, if we can get this .bat working and the GDAL wheel is included, it's a pretty easy install process for folks.

**Comment by ColinM9991** [comment]
> The reason I have a separate requirements_win.txt right now is macOS uses gdal 3.9.0 which is what homebrew installs and there doesn't appear to be a way to install a specific version via homebrew. On Windows, the latest published on cgohlke's repo is 3.8.4, so we're stuck using two different versions. I suppose someone could compile the latest version for Windows and we include that in /Utils/win....

You could consolidate both requirements files down to a single one by adding the following

```
gdal==3.9.0; sys_platform == 'darwin'
```

This will conditionally install gdal on mac. It does mean that we'll need to change `darwin` to target another platform name is Apple get around to doing that, but this would be part of the standard requirements versioning and maintenance.

> First idea though, why don't we just include the Python wheels for scikit-fmm? This prevents users from having to install the Microsoft Visual C++ Redistributable runtime libraries which is like a 6GB install. We're going to want to pin the dependencies anyways and it's only 65 KB. We could do the same for GDAL-3.8.4-cp312-cp312-win_amd64.whl (32MB) so it's one less step of having to download it. Both files can go in the Utils/win folder just to simplify things.

This might sound wacky - I am a C# developer so bear with me. This depends on whether our intention is for people to run Ortho4XP locally or develop it. What I mean by this is, we could add a GitHub workflow that uses something like PyInstaller that bundles the Ortho4XP scripts into a single self-contained package.

This then allows users to pick a release in GitHub and pull down the binaries without having to worry too much with dev dependencies. Meanwhile, developers and maintainers can setup their local environment with the required dev dependencies.

I know goes against the idea and conventions of Python since we're now effectively binarizing the scripts. Though it does lower the entry bar for non-developers who wish to run the latest version of Ortho4XP.

**Comment by w8sl** [comment]
Alternatively add pip install gdal==$(gdal-config --version) to the macOS install script or use python3.12 -m venv --system-site-packages. Same requirements.txt will then work in venv on all systems.

**Comment by shred86** [comment]
I completely forgot about PyInstaller - I think that's a perfect idea. I would guess that most users aren't going to be actively modifying Ortho4XP and would prefer to have something that's easy to setup. I wonder if there's a reason why PyInstaller isn't already being used. I just tried to package Ortho4XP using:

`pyinstaller --paths=src --hidden-import=PIL Ortho4XP.py`

And I get `Missing  ../Utils directory, check your install. Exiting.` when trying to launch the executable in `dist`. I tried passing `Utils` as a source too but no luck.

It's about a 280MB sized program with everything wrapped up.

**Comment by ColinM9991** [comment]
I'm playing around with `pyinstaller` now via the following

`pyinstaller --paths=src --add-data="Utils;Utils" --hidden-import=PIL Ortho4XP.py`

Since I'm using a venv, I'm receiving
`ModuleNotFoundError: No module named 'numpy'`

I'd imagine users would also receive this error without installing the dependencies. Trying to see if I can work around this by bundling dependencies with the binary.

I know Oscar's v1.3 Release for Windows is bundled but I'm not sure what process he used for doing this.

**Comment by ColinM9991** [comment]
Excuse the spam. Try this

```
pyinstaller --paths=src --contents-directory="." --add-data="Utils;Utils" --hidden-import=PIL Ortho4XP.py
```

**Comment by shred86** [comment]
Still getting the same error with that. I came across [this post](https://stackoverflow.com/questions/7674790/bundling-data-files-with-pyinstaller-onefile/13790741#13790741) and it seems to be a common issue folks run into with pyinstaller.

Implementing this, it seems to be working now! All of the folders are created in the default pyinstaller `_internal` folder but I'm sure that's something we can fix (so all your Ortho4XP stuff isn't intermixed with the dependencies).

This is looking very promising! Great idea. I haven't used pyinstaller before but it seems fairly straight forward so far.

**Comment by shred86** [comment]
Made some progress on this and just changed the latest to my [dev branch](https://github.com/shred86/Ortho4XP/tree/dev). Still obviously a work in progress but if you want to compile it yourself, you can clone that dev branch and use the following:

```
pyinstaller --noconfirm \
    --paths=src \
    --add-data="./Utils:./Utils" \
    --add-data="./Extents:./Extents" \
    --add-data="./Filters:./Filters" \
    --add-data="./Licence:./Licence" \
    --add-data="./Patches:./Patches" \
    --add-data="./Previews:./Previews" \
    --add-data="./Providers:./Providers" \
    --hidden-import=PIL \
    Ortho4XP.py
```

I've also uploaded it to Google Drive which you can download using [this link](https://drive.google.com/file/d/17op6AsXUyIoQjx0bvKaHtLEQZdW_DGaJ/view?usp=share_link) if anyone wants to help test it out.

I'm going to probably close this PR soon since I've opened an [issue](https://github.com/oscarpilote/Ortho4XP/issues/262) proposing the idea I mentioned above for @oscarpilote to hopefully see when he has time.

I opened up [Discussions](https://github.com/shred86/Ortho4XP/discussions) on my forked repo so I think that would be a good place for us to discuss any further development/documentation for now. Appreciate the help and discussions so far!

**Pull request closed without merging**
